src/filesystem/SDL_filesystem.c

```diff
@@ -0,0 +1,242 @@
+  including commercial applications, and to alter it and redistribute it
+  freely, subject to the following restrictions:
+
+  1. The origin of this software must not be misrepresented; you must not
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+
+#include "SDL_internal.h"
+#include "SDL_sysfilesystem.h"
+
+#ifdef SDL_PLATFORM_WINDOWS
+    #define PLATFORM_PATH_SEPARATOR "\\"
+    #define PLATFORM_ROOT_PATH ""
+#else
+    #define PLATFORM_PATH_SEPARATOR  "/"
+    #define PLATFORM_ROOT_PATH "/"
+#endif
+
+
+SDL_PropertiesID SDL_GetFSProperties(SDL_FSops *context)
+{
+    if (!context) {
+        SDL_InvalidParamError("context");
+        return 0;
+    }
+
+    if (context->props == 0) {
+        context->props = SDL_CreateProperties();
+    }
+    return context->props;
+}
+
+SDL_RWops *SDL_FSopen(SDL_FSops *fs, const char *path, const char *mode)
+{
+    SDL_RWops *retval = NULL;
+    if (!fs) {
+        SDL_InvalidParamError("fs");
+    } else if (!path) {
+        SDL_InvalidParamError("path");
+    } else if (!mode) {
+        SDL_InvalidParamError("mode");
+    } else {
+        retval = fs->open(fs, path, mode);
+    }
+    return retval;
+}
+
+int SDL_FSenumerate(SDL_FSops *fs, const char *path, SDL_EnumerateCallback cb, void *userdata)
+{
+    if (!fs) {
+        return SDL_InvalidParamError("fs");
+    } else if (!path) {
+        SDL_InvalidParamError("path");
+    } else if (!cb) {
+        return SDL_InvalidParamError("cb");
+    }
+    return fs->enumerate(fs, path, cb, userdata);
+}
+
+int SDL_FSremove(SDL_FSops *fs, const char *path)
+{
+    if (!fs) {
+        return SDL_InvalidParamError("fs");
+    } else if (!path) {
+        return SDL_InvalidParamError("path");
+    }
+    return fs->remove(fs, path);
+}
+
+int SDL_FSmkdir(SDL_FSops *fs, const char *path)
+{
+    if (!fs) {
+        return SDL_InvalidParamError("fs");
+    } else if (!path) {
+        return SDL_InvalidParamError("path");
+    }
+    return fs->mkdir(fs, path);
+}
+
+int SDL_FSstat(SDL_FSops *fs, const char *path, SDL_Stat *stat)
+{
+    if (!fs) {
+        return SDL_InvalidParamError("fs");
+    } else if (!path) {
+        return SDL_InvalidParamError("path");
+    } else if (!stat) {
+        return SDL_InvalidParamError("stat");
+    }
+    return fs->stat(fs, path, stat);
+}
+
+const char *SDL_GetFilePathSeparator(void)
+{
+    return PLATFORM_PATH_SEPARATOR;
+}
+
+void SDL_FileTimeToWindows(Sint64 ftime, Uint32 *low, Uint32 *high)
+{
+// !!! FIXME: check this math.
+    const Uint64 epoch_convert = 0x019DB1DED53E8000ull;
+    const Uint64 cvt = (Uint64) ((ftime + epoch_convert) * 10000000);
```

Comment:
I think it needs to be:
```suggestion
    const Uint64 delta_1601_epoch_s = 11644473600; /* [seconds] (seconds between 1/1/1601 and 1/1/1970)*/
    const Uint64 cvt = (delta_1601_to_epoch_s + ftime) * 10000000ull; /* [100ns] */
```

```python
>>> import datetime
>>> int((datetime.datetime(year=1970, month=1, day=1) - datetime.datetime(year=1601, month=1, day=1)).total_seconds())
11644473600
```

Reply:
0x019DB1DED53E8000 is 116444736000000000, or 11644473600 times 10000000, so I probably should have had `(ftime * 10000000) + epoch_convert` (maybe?) but your patch is more clear than my code in any case.

Reply:
Is epoch is in seconds right? No milliseconds. (I didn't read or try your code in detail yet)
You need to add seconds with seconds, and multiples of 100ns with multiples of 100ns.
So you can do `ftime * 10000000 +  116444736000000000` as well.
` 116444736000000000` is an offset, not a multiplier factor.

Reply:
Resolved in 6ed943fba7394c2ce84169f157c975d3c19294b5.